src/main/drivers/bus_i2c.c

```diff
@@ -0,0 +1,441 @@
+            if (subaddress_sent || (state->writing)) {
+                if (final_stop)
+                    I2C_GenerateSTOP(I2Cx, ENABLE);                     // program the Stop
+                else
+                    I2C_GenerateSTART(I2Cx, ENABLE);                    // program a rep start
+                index++;                                                // to show that the job is complete
+            } else {                                                    // We need to send a subaddress
+                I2C_GenerateSTART(I2Cx, ENABLE);                        // program the repeated Start
+                subaddress_sent = 1;                                    // this is set back to zero upon completion of the current task
+            }
+        }
+        // we must wait for the start to clear, otherwise we get constant BTF
+        while (I2Cx->CR1 & 0x0100) { ; }
+    } else if (SReg_1 & I2C_SR1_RXNE) {                                 // Byte received - EV7
+        state->read_p[index++] = (uint8_t)I2Cx->DR;
+        if (state->bytes == (index + 3))
+            I2C_ITConfig(I2Cx, I2C_IT_BUF, DISABLE);                    // disable TXE to allow the buffer to flush so we can get an EV7_2
+        if (state->bytes == index)                                             // We have completed a final EV7
+            index++;                                                    // to show job is complete
+    } else if (SReg_1 & I2C_SR1_TXE) {                                  // Byte transmitted EV8 / EV8_1
+        if (index != -1) {                                              // we dont have a subaddress to send
+            I2Cx->DR = state->write_p[index++];
+            if (state->bytes == index)                                         // we have sent all the data
+                I2C_ITConfig(I2Cx, I2C_IT_BUF, DISABLE);                // disable TXE to allow the buffer to flush
+        } else {
+            index++;
+            I2Cx->DR = state->reg;                                             // send the subaddress
+            if (state->reading || !(state->bytes))                                      // if receiving or sending 0 bytes, flush now
+                I2C_ITConfig(I2Cx, I2C_IT_BUF, DISABLE);                // disable TXE to allow the buffer to flush
+        }
+    }
+    if (index == state->bytes + 1) {                                           // we have completed the current job
+        subaddress_sent = 0;                                            // reset this here
+        if (final_stop)                                                 // If there is a final stop and no more jobs, bus is inactive, disable interrupts to prevent BTF
+            I2C_ITConfig(I2Cx, I2C_IT_EVT | I2C_IT_ERR, DISABLE);       // Disable EVT and ERR interrupts while bus inactive
+        state->busy = 0;
+    }
+}
+
+void i2cInit(I2CDevice index) 
+{
+    i2cDevice_t device;
```

Comment:
You are mixing `device` as index and `device` as pointer to device descriptor. Try to use consistent naming